### ISSUE TITLE
selecting a file/dir from directory listing with symlinking

### ISSUE THREAD

**Issue by jfouret** [common-workflow-language/cwltool#1391]
## Expected Behavior

In an ExpressionTool  and when selecting a file or subdirectory from an input directory,
I want this file not to be copied but rather symlinked.

Is there a way to do that ?

This feature would be really useful when files are big. Especially when, in a workflow the expressionTool output is an intermediary file which is not kept in final output.

## Actual Behavior

I did not find any way to implement the behavior I am looking for. The selected file is copied in my workflows. 
Below an example of an expressionTool to select a file from a directory.

## Workflow Code
```
cwlVersion: v1.0
class: ExpressionTool
id: fileFromDir
requirements:
  InlineJavascriptRequirement: {}
inputs:
  dir:
    type: Directory
  prefix:
    type: string
    default: ''
  basename:
    type: string
  suffix:
    type: string
outputs:
  xfile:
    type: 'File'
expression: >
  ${
    var targetFileName=inputs.prefix+inputs.basename+inputs.suffix;
    for (var i in inputs.dir.listing){
      var itemFile=inputs.dir.listing[i];
      if (itemFile.basename == targetFileName){
        var targetFile=itemFile;
        break;
      }
    }
    return {'xfile':targetFile};
  }
```

## Your Environment
* cwltool version: 3.0.20200709181526

**Comment by mr-c** [comment]
Hello @jfouret ; can you send a workflow that uses this ExpressionTool so we can see the file copying you are experiencing?

My understanding is that cwltool only copies when `writable: true` is specified via `InitialWorkDirRequirement` & [`Dirent`](https://www.commonwl.org/v1.0/CommandLineTool.html#Dirent)

**Comment by jfouret** [comment]
Hi, 

I understand that the default is `writable:false`. But I think that this option is not pertinent here. In fact the file, `xfile`, returned by `fileFromDir` is considered as an output and is therefore copied. Though it is totally expected from the current specifications, CWL interpreter has no way to know that it is file already present in a input directory and thus that a symlink is possible (should be preferred).

Below is the workflow I am using. The principle is to give as input a directory with all fastq files `fastqDir` (some paired-end and other single end) then based on suffix patterns for read 1 and read 2 (`R1_suffix` and `R2_suffix`), the ExpressionTool `fastqdir2samplenames` outputs two list of strings (`sampleNames_pe`, the basenames of samples with paired-end reads and `sampleNames_sr`, the basenames of samples with single-read design ).

Then the ExpressionTool `fileFromDir` is used to extract the fastq files from the input directory `fastqDir` for the file to be given as input to different tools (here classical) : `cutadapt`, `fastq_screen` and `fastqc`.

In this workflow the all fastq files are copied whereas it would be more efficient if they were symlinked.

I was not able to think of a way to make them symlinked based on the current specifications (I have also looked at 1.1 and 1.2). Maybe I took the problem from the wrong side or maybe there is simply no way with the current specifications (open question) ?

Below is an example of workflow with the `fastqdir2samplenames` to fully understand it.

## workflow

```
class: Workflow
cwlVersion: v1.0
id: ngs_primary_qc
label: ngs_primary_qc
inputs:
  - id: genomesDir
    type: Directory
  - id: conf
    type: File
  - id: min_length
    type: int
    default: 45
  - id: ncpus_per_job
    type: int
    default: 4
  - id: R1_suffix
    type: string
    default: _R1.fastq.gz
  - id: R2_suffix
    type: string
    default: _R2.fastq.gz
  - id: fastqDir
    type: Directory
outputs:
  - id: qc_report
    outputSource:
      - multiqc/qc_report
    type: File
  - id: trimmed_reads2
    outputSource:
      - cutadapt_pe/trimmed_reads2
    type: 'File[]'
  - id: trimmed_reads1
    outputSource:
      - cutadapt_pe/trimmed_reads1
    type: 'File[]'
  - id: trimmed_reads1_sr
    outputSource:
      - cutadapt_sr/trimmed_reads1
    type: 'File[]'
steps:
  - id: multiqc
    in: 
      - id: dirs
        linkMerge: merge_flattened
        source:
          - fastqc/outdir
          - fastq_screen/outdir
    out:
      - id: qc_report
    run: tools/multiqc/tool.cwl
  - id: fastqc
    in:
      - id: ncpus
        source: ncpus_per_job
      - id: reads
        linkMerge: merge_flattened
        source:
          - pe_get_R1/xfile
          - pe_get_R2/xfile
          - sr_get_R1/xfile
          - cutadapt_pe/trimmed_reads2
          - cutadapt_pe/trimmed_reads1
          - cutadapt_sr/trimmed_reads1
    out:
      - id: outdir
    run: tools/fastqc/tool.cwl
    scatter:
      - reads
    scatterMethod: nested_crossproduct
  - id: cutadapt_pe
    in:
      - id: ncpus
        source: ncpus_per_job
      - id: min_length
        source: min_length
      - id: reads1
        source: pe_get_R1/xfile
      - id: reads2
        source: pe_get_R2/xfile
    out:
      - id: trimmed_reads1
      - id: trimmed_reads2
    run: tools/cutadapt/tool.cwl
    label: cutadapt_pe
    scatter:
      - reads1
      - reads2
    scatterMethod: dotproduct
  - id: cutadapt_sr
    in:
      - id: ncpus
        source: ncpus_per_job
      - id: min_length
        source: min_length
      - id: reads1
        source: sr_get_R1/xfile
    out:
      - id: trimmed_reads1
    run: tools/cutadapt/tool_sr.cwl
    label: cutadapt_sr
    scatter:
      - reads1
    scatterMethod: dotproduct
  - id: fastq_screen
    in:
      - id: ncpus
        source: ncpus_per_job
      - id: conf
        source: conf
      - id: genomesDir
        source: genomesDir
      - id: reads
        linkMerge: merge_flattened
        source:
          - cutadapt_pe/trimmed_reads1
          - cutadapt_sr/trimmed_reads1
    out:
      - id: outdir
    run: tools/fastq_screen/tool.cwl
    scatter:
      - reads
    scatterMethod: dotproduct
  - id: fastqdir2samplenames
    in:
      - id: R1
        source: R1_suffix
      - id: R2
        source: R2_suffix
      - id: fastqDir
        source: fastqDir
    out:
      - id: sampleNames_pe
      - id: sampleNames_sr
    run: tools/fastqdir2samplenames/tool.cwl
  - id: sr_get_R1
    in:
      - id: dir
        source: fastqDir
      - id: basename
        source: fastqdir2samplenames/sampleNames_sr
      - id: suffix
        source: R1_suffix
    out:
      - id: xfile
    run: tools/fileFromDir/tool.cwl
    scatter:
      - basename
    scatterMethod: dotproduct
  - id: pe_get_R1
    in:
      - id: dir
        source: fastqDir
      - id: basename
        source: fastqdir2samplenames/sampleNames_pe
      - id: suffix
        source: R1_suffix
    out:
      - id: xfile
    run: tools/fileFromDir/tool.cwl
    scatter:
      - basename
    scatterMethod: dotproduct
  - id: pe_get_R2
    in:
      - id: dir
        source: fastqDir
      - id: basename
        source: fastqdir2samplenames/sampleNames_pe
      - id: suffix
        source: R2_suffix
    out:
      - id: xfile
    run: tools/fileFromDir/tool.cwl
    scatter:
      - basename
    scatterMethod: dotproduct
requirements:
  - class: ScatterFeatureRequirement
  - class: MultipleInputFeatureRequirement
```
## expressionTool `fastqdir2samplenames`

```
cwlVersion: v1.0
class: ExpressionTool
id: fastqdir2samplenames
requirements:
  InlineJavascriptRequirement: {}
inputs:
  prefix:
    type: string
    default: ''
  fastqDir:
    type: Directory
  R1:
    type: string
    default: "_R1.fastq.gz"
  R2:
    type: string
    default: "_R2.fastq.gz"
outputs:
  sampleNames_sr:
    type: 'string[]'
  sampleNames_pe:
    type: 'string[]'
expression: >
  ${
    var R1Names=[];
    var R2Names=[];
    var re_R1=new RegExp(inputs.R1+'$');
    var re_R2=new RegExp(inputs.R2+'$');
    var re_prefix=new RegExp('^'+inputs.prefix);
    for (var i in inputs.fastqDir.listing){
      var itemFile=inputs.fastqDir.listing[i];
      if (itemFile.basename.startsWith(inputs.prefix)){
        if (itemFile.basename.endsWith(inputs.R1)){
          R1Names.push(itemFile.basename.replace(re_R1,'').replace(re_prefix,''));
        }
        else if (itemFile.basename.endsWith(inputs.R2)){
          R2Names.push(itemFile.basename.replace(re_R2,'').replace(re_prefix,''));
        }
      }
    }
    var peNames=[];
    var srNames=[];
    for (var i in R1Names){
      var R1Name=R1Names[i];
      if (R2Names.includes(R1Name)){
        peNames.push(R1Name);
      }else{
        srNames.push(R1Name);
      }
    }
    return {
      'sampleNames_pe':peNames,
      'sampleNames_sr':srNames
    };
  }

```

**Comment by mr-c** [comment]
Are you sure that a symlink is not being used?

Here's how I tested this

```
$ cwltool --enable-ext file-picker-wf.cwl --my_dir tests
INFO /home/michael/cwltool/env3.8/bin/cwltool 3.0.20201203182617
INFO Resolved 'file-picker-wf.cwl' to 'file:///home/michael/cwltool/file-picker-wf.cwl'
INFO [workflow ] start
INFO [workflow ] starting step first
INFO [step first] start
INFO [step first] completed success
INFO [workflow ] starting step second
INFO [step second] start
INFO [job second] /tmp/phe0whb5$ ls \
    -la \
    /tmp/86fw6ri1/stg8c71d9eb-0875-4e77-9e94-0308e8a4ee09/echo-position-expr-job.yml > /tmp/phe0whb5/1450fcd9d036f71a53816b53b8154c67fa32be37
INFO [job second] completed success
INFO [step second] completed success
INFO [workflow ] completed success
{
    "listing": {
        "location": "file:///home/michael/cwltool/1450fcd9d036f71a53816b53b8154c67fa32be37",
        "basename": "1450fcd9d036f71a53816b53b8154c67fa32be37",
        "class": "File",
        "checksum": "sha1$a4ad5a0bab2ee81f463c92b4b62d7cf8d76e2074",
        "size": 184,
        "path": "/home/michael/cwltool/1450fcd9d036f71a53816b53b8154c67fa32be37"
    }
}
INFO Final process status is success
$ cat 1450fcd9d036f71a53816b53b8154c67fa32be37
lrwxrwxrwx 1 michael michael 54 Dec 16 13:22 /tmp/86fw6ri1/stg8c71d9eb-0875-4e77-9e94-0308e8a4ee09/echo-position-expr-job.yml -> /home/michael/cwltool/tests/echo-position-expr-job.yml
```
So it is a symlink, not a copy

here is `file-picker-wf.cwl`
```
cwlVersion: v1.0
class: Workflow

inputs:
 my_dir: Directory

steps:
  first:
    run:
      class: ExpressionTool
      requirements:
        InlineJavascriptRequirement: {}
      inputs:
        dir:
          type: Directory
      outputs:
        xfile:
          type: File
      expression: '${return {"xfile":inputs.dir.listing[0]};}'
    in:
      dir: my_dir
    out: [ xfile]

  second:
    run:
      class: CommandLineTool
      inputs:
       a_file: File
      baseCommand: [ls, -la]
      arguments: [ $(inputs.a_file) ]
      outputs:
        listing: stdout
    in:
      a_file: first/xfile
    out: [ listing ]

outputs:
 listing:
   type: File
   outputSource: second/listing


$namespaces:
  cwltool: "http://commonwl.org/cwltool#"
hints:
  cwltool:LoadListingRequirement:
    loadListing: shallow_listing
```

**Comment by jfouret** [comment]
Thank you very much for your reactivity.

My bad, everything is fine with this implementation in cwltool.

Since the problem I had related to #888 I have temporarely replaced `cwltool` cmdline with `toil-cwl-runner`.

Using cwltool and while keeping temporary file we ccan see that everything is fine:

```
$cwltool --parallel --force-docker-pull --verbose --no-match-user --on-error 'continue' --tmpdir-prefix tmp --cachedir cache --leave-tmpdir --timestamps --outdir outdir file-picker-wf.cwl --my_dir test_dir

[...]

$ tree .
.
├── cache
│   ├── 0e647ac658d6b4319997e3bdbb3e5a33
│   │   └── 413bfea21ae37cc658b007d2fd2b80b4d5dd3db0
│   └── 0e647ac658d6b4319997e3bdbb3e5a33.status
├── file-picker-wf.cwl
├── outdir
│   └── 413bfea21ae37cc658b007d2fd2b80b4d5dd3db0
├── test_dir
│   ├── hello1.txt
│   └── hello2.txt
└── tmp

5 directories, 6 files

```
The file `hello1.txt` is not copied; nor in `cache` nor in `tmp`.

However the problem stands with `toil-cwl-runner`:

```
$toil-cwl-runner --jobStore file:cache/test --strict-memory-limit --force-docker-pull --logDebug --clean never --outdir outdir file-picker-wf.cwl --my_dir test_dir

[...]

$ tree .

.
├── cache
│   └── test
│       ├── files
│       │   ├── for-job
│       │   │   ├── kind-CWLWorkflow
│       │   │   │   └── instance-f0yieq5v
│       │   │   └── kind-file_home_users_jfouret_toil_test_file-picker-wf.cwl_first_20f90064-e089-49f8-b671-2b61809f0bf9
│       │   │       └── instance-wqh50svk
│       │   │           ├── file-c7rkl6um
│       │   │           │   └── hello1.txt
│       │   │           └── file-zktxmvep
│       │   │               └── 65588ba8669eb87edeb7befb2706bef493704366 -> /home/users/jfouret/toil_test/outdir/65588ba8669eb87edeb7befb2706bef493704366
│       │   ├── no-job
│       │   │   ├── file-3l6j87kh
│       │   │   ├── file-d197eyo7
│       │   │   │   └── stream
│       │   │   ├── file-qyhetvsm
│       │   │   ├── file-sc4qb0xf
│       │   │   │   └── stream
│       │   │   └── file-z9hgcs_j
│       │   └── shared
│       │       ├── config.pickle
│       │       ├── environment.pickle
│       │       ├── pid.log
│       │       ├── rootJobReturnValue
│       │       ├── rootJobStoreID
│       │       └── succeeded.log
│       ├── jobs
│       │   ├── kind-CWLWorkflow
│       │   ├── kind-ResolveIndirect
│       │   ├── kind-file_home_users_jfouret_toil_test_file-picker-wf.cwl_first_20f90064-e089-49f8-b671-2b61809f0bf9
│       │   └── kind-file_home_users_jfouret_toil_test_file-picker-wf.cwl_second_50e162d9-d940-426c-b499-3f1ab7e8218b
│       └── stats
│           ├── stats8p8nzd5o.new
│           ├── stats8pz85nbi.new
│           ├── statsacbvpvk8.new
│           ├── statsnjpn2rro
│           └── statsvrbu1oen.new
├── file-picker-wf.cwl
├── outdir
│   ├── 2093ea3651180b15e0b70374fab093583577c459
│   └── 65588ba8669eb87edeb7befb2706bef493704366
├── test_dir
│   ├── hello1.txt
│   └── hello2.txt
└── tmp

26 directories, 20 files
```
The problem is that the `hello1.txt` file is copied in the temporary directory (jobstore) `cache/test/files/for-job/kind-file_home_users_jfouret_toil_test_file-picker-wf.cwl_first_20f90064-e089-49f8-b671-2b61809f0bf9/instance-wqh50svk/file-c7rkl6um/hello1.txt` rather than being symlinked.

I did not see that #888 had been fixed. I will try it.

I am going to create an issue for toil and reference this post.

Thanks,

**Comment by mr-c** [comment]
Thank you @jfouret for the update and your persistence! I'll close this issue for now. I hope the fix for #888 works for you :-)
